### PR TITLE
feat: Add random image cron

### PR DESCRIPTION
**Pull request by KanteTran**

### CODE REVIEW

File: bot/telegram/telegram_bot.go [hunk 1]
```diff
@@ -44,6 +49,40 @@ func (b *Bot) Serve() error {
 
 	updates := bot.GetUpdatesChan(u)
 
+	location, err := time.LoadLocation("Asia/Ho_Chi_Minh")
```
Comment (hvn2706) [inline bot/telegram/telegram_bot.go]: Should plit into different function, add "Asia/Ho_Chi_Minh" inside [constant package](https://github.com/hvn2706/jsin/blob/67c210da41c22fa2ade3a028a9915b7ad924780c/pkg/constants/constant.go)

File: bot/message_handler/bot.go [hunk 2]
```diff
@@ -13,7 +13,6 @@ import (
 
 type IMessageHandler interface {
 	HandleMessage(ctx context.Context, message string) (*MessageDTO, error)
-
```
Comment (hvn2706) [inline bot/message_handler/bot.go:17]: Reverse unecessary changes please, keep in mind the lint check

File: bot/telegram/telegram_bot.go [hunk 3]
```diff
@@ -44,6 +49,40 @@ func (b *Bot) Serve() error {
...
+		log.Fatalf("Failed to load location: %v", err)
+	}
+
+	c := cron.New(cron.WithLocation(location))
+
+	_, err = c.AddFunc(b.cfg.CronJobRandomImage, func() {
+		for _, idChat := range b.cfg.WhileLists {
+			idNeedToSend, err := strconv.ParseInt(idChat, 10, 64)
+			if err != nil {
+				log.Printf("Invalid chat ID: %v", err)
```
Comment (hvn2706) [inline bot/telegram/telegram_bot.go]: use our internal logger package please!

File: bot/telegram/telegram_bot.go [hunk 4]
```diff
@@ -44,6 +49,40 @@ func (b *Bot) Serve() error {
...
+			if err != nil {
+				log.Printf("Failed to send image to chat ID %d: %v", idNeedToSend, err)
+			} else {
+				log.Printf("Daily image sent successfully to chat ID %d.", idNeedToSend)
+			}
+		}
+	})
+
+	if err != nil {
+		log.Fatalf("Error scheduling cron job: %v", err)
```
Comment (hvn2706) [inline bot/telegram/telegram_bot.go]: Do not fatal the service!

File: cmd/main.go [hunk 5]
```diff
@@ -15,16 +14,13 @@ import (
 )
 
 func main() {
-	// 1. Load config
```
Comment (hvn2706) [inline cmd/main.go]: I don't think we should delete comments like this even though we think that it is unnecessary and obvious

File: config.tmp.yml [hunk 6]
```diff
@@ -31,3 +31,5 @@ telegram_bot:
   debug: false
   offset: 0
   timeout: 60
+  white_lists: ["-1002437731689"] # list id chat in tele will receive daily random image
```
Comment (hvn2706) [inline config.tmp.yml]: Keep in mind that we don't use real value here!

**Review by hvn2706 — CHANGES_REQUESTED** [review]

File: bot/telegram/send_image.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+	message, err := b.bot.Send(
+		tgbotapi.PhotoConfig{
+			BaseFile: tgbotapi.BaseFile{
+				BaseChat: tgbotapi.BaseChat{
+					ChatID: chatID,
+				},
+				File: file,
+			},
+			Caption: "🌅 Have a great day!",
```
Comment (hvn2706) [inline bot/telegram/send_image.go]: Please remember that we put these strings like magic key words or numbers in a constant value!

File: bot/telegram/send_image.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+		logger.Errorf("===== Send image failed: %+v", err.Error())
+		return err
+	}
+
+	logger.Infof("Image sent, message id: %d", message.MessageID)
+	return nil
+}
+
+func (b *Bot) SendImageRandomDaily(chatID int64) error {
+	currentTime := time.Now().Format("12-12-2024")
```
Comment (hvn2706) [inline bot/telegram/send_image.go]: magic number!

File: bot/message_handler/bot.go (hunk 2 above)
Comment (KanteTran) [reply bot/message_handler/bot.go:17]: Agree with you, bro

File: bot/telegram/send_image.go (hunk 8 above)
Comment (KanteTran) [reply bot/telegram/send_image.go]: Good catch! Would you prefer I use a constant, or make the format configurable via a function parameter?

File: bot/telegram/telegram_bot.go (hunk 3 above)
Comment (KanteTran) [reply bot/telegram/telegram_bot.go]: I updated. Please resolve

File: bot/telegram/telegram_bot.go (hunk 4 above)
Comment (KanteTran) [reply bot/telegram/telegram_bot.go]: I updated. Please resolve

File: config.tmp.yml (hunk 6 above)
Comment (KanteTran) [reply config.tmp.yml]: Agree

File: bot/telegram/send_image.go (hunk 7 above)
Comment (KanteTran) [reply bot/telegram/send_image.go]: Updated. Thanks

File: bot/telegram/telegram_bot.go (hunk 1 above)
Comment (KanteTran) [reply bot/telegram/telegram_bot.go]: Good catch. Thanks

File: bot/telegram/send_image.go (hunk 8 above)
Comment (hvn2706) [reply bot/telegram/send_image.go]: just make it as a constant

File: config.tmp.yml [hunk 9]
```diff
@@ -31,7 +31,9 @@ telegram_bot:
   debug: false
   offset: 0
   timeout: 60
+  white_lists: ["chatIds"] # list chatIds will receive daily random image
+  cron_job_random_image: "* * * * * "
 
-# you can explain your image type here
```
Comment (hvn2706) [inline config.tmp.yml:35]: please do not delete comments used for explaination

File: bot/telegram/telegram_bot.go [hunk 10]
```diff
@@ -43,6 +48,40 @@ func (b *Bot) Serve() error {
...
+		}
+	})
+
+	if err != nil {
+		logger.Errorf("Error scheduling cron job: %v", err)
+	}
+
+	c.Start()
+
+	if err != nil {
```
Comment (hvn2706) [inline bot/telegram/telegram_bot.go]: please log the error

File: bot/telegram/telegram_bot.go [hunk 11]
```diff
@@ -43,6 +48,40 @@ func (b *Bot) Serve() error {
...
+			if err != nil {
+				logger.Errorf("Failed to send image to chat ID %d: %v", idNeedToSend, err)
+			} else {
+				logger.Infof("Daily image sent successfully to chat ID %d.", idNeedToSend)
+			}
+		}
+	})
+
+	if err != nil {
+		logger.Errorf("Error scheduling cron job: %v", err)
```
Comment (hvn2706) [inline bot/telegram/telegram_bot.go]: should it return err here to finish the function?

Comment (KanteTran) [reply bot/telegram/telegram_bot.go]: If you don't want the cron job error to affect the main bot execution flow, then there's no need to return after logging the error.

File: bot/telegram/send_image.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+	if generateContent != nil && generateContent.Object != nil {
+		err = b.SendImageByObject(chatID, *generateContent.Object)
+
+		if err != nil {
+			logger.Errorf("===== Send image failed: %+v", err.Error())
+			return err
+		}
+	}
+
+	logger.Infof("Daily good morning message and image sent successfully.")
```
Comment (hvn2706) [inline bot/telegram/send_image.go]: this is not going to be daily message, isn't it? And let's embed more info inside this log for more information!

File: bot/telegram/telegram_bot.go [hunk 13]
```diff
@@ -43,6 +47,36 @@ func (b *Bot) Serve() error {
 
 	updates := bot.GetUpdatesChan(u)
 
+	location, err := time.LoadLocation(constants.TimeZone)
```
Comment (hvn2706) [inline bot/telegram/telegram_bot.go]: again, split this into different function please

File: pkg/storage/cronjob_storage.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+	cronJob string,
+	jobType string,
+) (int64, error) {
+	newJob := map[string]interface{}{
+		"chat_id":  ctx.Value("chatID"),
+		"cron_job": cronJob,
+		"type":     jobType,
+	}
+
+	result := i.gdb.DB().Table("cron_job").Create(newJob)
```
Comment (hvn2706) [inline pkg/storage/cronjob_storage.go]: you should create a model for this table, reference: 

https://github.com/hvn2706/jsin/blob/main/model/image.go
https://github.com/hvn2706/jsin/blob/main/pkg/storage/image_storage.go#L28-L48

then you can remove some magic keyword in the code

File: pkg/storage/cronjob_storage.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	insertedID := result.RowsAffected
+	return insertedID, nil
+}
+
+func (i *CronJobStorageImpl) ListCronJobDaily(ctx context.Context) ([]CronJob, error) {
+	var cronJobs []CronJob
+
+	result := i.gdb.DB().Table("cron_job").
```
Comment (hvn2706) [inline pkg/storage/cronjob_storage.go]: same here!

File: config.tmp.yml [hunk 16]
```diff
@@ -35,3 +35,6 @@ telegram_bot:
 # you can explain your image type here
 help_command_content: |
   j.sin to generate content
+
+create_cronjob_command_content: |
+    j.sin will send you random gift
```
Comment (hvn2706) [inline config.tmp.yml]: let's use 2 spaces tab instead of 4 here

**Review by hvn2706 — COMMENTED** [review]
@KanteTran Nice, the design looks more elastic! Can you think about a way to limit number of cronjobs a channel/user can have to only 1 or 2? because I don't think we want duplicate job that do the same think.

What do you think?

**Comment by KanteTran** [comment]
> @KanteTran Nice, the design looks more elastic! Can you think about a way to limit number of cronjobs a channel/user can have to only 1 or 2? because I don't think we want duplicate job that do the same think.
> 
> What do you think?

We have a de

> @KanteTran Nice, the design looks more elastic! Can you think about a way to limit number of cronjobs a channel/user can have to only 1 or 2? because I don't think we want duplicate job that do the same think.
> 
> What do you think?

We have a separate service to handle reading cronjobs, and we can process the logic independently there.

In my opinion, limiting the daily random cronjob is unnecessary because the number of cronjobs that can run is already limited.

If you want to impose limits, it would be better to control who is allowed to create cronjobs, I think that’s a more effective approach.

Lastly, I believe this should be a separate commit. It’s better not to include too many ideas in a single commit, bro. @hvn2706

File: pkg/storage/cronjob_storage.go (hunk 14 above)
Comment (KanteTran) [reply pkg/storage/cronjob_storage.go]: Good catch

**Pull request merged**